### PR TITLE
Refactor code index page

### PR DESCRIPTION
**Pull request by tbille**
# Summary

Refactor code that was only in the django view to make it easier when adding flask

Add embed param to remove a few API calls that we are doing

# QA

## I know I am not using the run script this is to make it easier to dev and qa

On this branch:
- `poetry`
- `poetry build`
- copy setup.py to the root of the project

On cn.ubuntu.com

- comment blog module and add this branch

```git
diff --git a/requirements.txt b/requirements.txt
index 1de469c..e4e20ad 100644
--- a/requirements.txt
+++ b/requirements.txt
@@ -1,5 +1,5 @@
 Django==2.2
-canonicalwebteam.blog==1.3.5
+# canonicalwebteam.blog==1.3.5
 canonicalwebteam.http==1.0.1
 canonicalwebteam.versioned-static==1.0.2
 canonicalwebteam.yaml-responses[django]==1.1.0
@@ -9,3 +9,5 @@ django-template-finder-view==0.3
 whitenoise==4.1.2
 talisker==0.14.2
 gunicorn[gevent]
+
+-e file:///home/tbmb/Documents/projects/blog-flask-extension#egg=canonicalwebteam.blog
```

- On a python 3 virtualenv
- `pip install -r requirements.txt`
- `yarn && yarn build`
- `python manage.py runserver`

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1 +1,5 @@
 **/__pycache__
+canonicalwebteam.blog.egg-info/
```
Comment (jpmartinspt) [inline .gitignore]: `*.egg-info/`

File: canonicalwebteam/blog/common_view_logic.py [hunk 2]
```diff
@@ -5,39 +5,104 @@
...
+class BlogViews:
+    def __init__(self, tag_ids, excluded_tags, blog_title, tag_name):
+        self.tag_ids = tag_ids
+        self.excluded_tags = excluded_tags
+        self.blog_title = blog_title
+        self.tag_name = tag_name
+
+    def get_index(self, page=1, category="", enable_upcoming=True):
+        category_id = ""
+        if category != "":
```
Comment (jpmartinspt) [inline canonicalwebteam/blog/common_view_logic.py]: `if category:` should be enough

File: canonicalwebteam/blog/common_view_logic.py [hunk 3]
```diff
@@ -5,39 +5,104 @@
...
-                resolved_group = api.get_group_by_id(group_id)
-                group_cache[group_id] = resolved_group
-            if first_item:
-                article["group"] = group_cache[group_id]
-                first_item = False
+        if (
+            "wp:term" in article["_embedded"]
+            and article["_embedded"]["wp:term"][3]
+        ):
+            article["group"] = article["_embedded"]["wp:term"][3][0]
```
Comment (jpmartinspt) [inline canonicalwebteam/blog/common_view_logic.py]: Might be worth documenting why we use this indexes.

Comment (tbille) [reply canonicalwebteam/blog/common_view_logic.py]: I agree

Comment (tbille) [reply canonicalwebteam/blog/common_view_logic.py]: I extracted all of it into "getters" functions

File: tests/test_wordpress_api.py [hunk 4]
```diff
@@ -95,6 +95,7 @@ def test_getting_24_articles_from_page_3(self, get):
             + "&categories="
             + "&exclude="
             + "&author="
+            + "&_embed"
```
Comment (b-m-f) [inline tests/test_wordpress_api.py]: Hm,

this one https://github.com/canonical-web-and-design/canonicalwebteam.blog/pull/55 should simplify testing quite a bit :)

Maybe you want to rebase when its merged?

**Comment by b-m-f** [comment]
Sweet. Thanks for starting this already :)

File: canonicalwebteam/blog/common_view_logic.py [hunk 5]
```diff
@@ -5,40 +5,151 @@
...
+    """Returns the featured media in the embedded response from wp.
+    embedded["wp:featuredmedia"]
+
+
+    :param embedded: The embedded dictionnary in the response
+    :returns: List of featuredmedia
+    """
+    if "wp:featuredmedia" in embedded:
+        return embedded["wp:featuredmedia"]
+    return []
```
Comment (jpmartinspt) [inline canonicalwebteam/blog/common_view_logic.py]: return embedded.get("wp:featuredmedia". [])

File: canonicalwebteam/blog/common_view_logic.py [hunk 6]
```diff
@@ -5,40 +5,151 @@
...
+    """Returns the author in the embedded response from wp.
+    embedded["author"]
+
+
+    :param embedded: The embedded dictionnary in the response
+    :returns: Dictionnary of author
+    """
+    if "author" in embedded:
+        return embedded["author"][0]
+    return {}
```
Comment (jpmartinspt) [inline canonicalwebteam/blog/common_view_logic.py]: return embedded.get("author", [{}])[0]?

Comment (tbille) [reply canonicalwebteam/blog/common_view_logic.py]: I am not a big fan of calling the element of a list directly after a function call.

Comment (tbille) [reply canonicalwebteam/blog/common_view_logic.py]:
```
    authors = embedded.get("author", [{}])
    return authors[0]
```

?

**Comment by jpmartinspt** [comment]
Needs the version bump on `setup.py` and `CHANGELOG`.

File: canonicalwebteam/blog/common_view_logic.py [hunk 7]
```diff
@@ -5,39 +5,145 @@
...
+
+    def get_index(self, page=1, category="", enable_upcoming=True):
+        category_id = ""
+        if category:
+            category_resolved = api.get_category_by_slug(category)
+            category_id = category_resolved["id"]
+
+        upcoming = []
+        featured_articles = []
+        featured_article_ids = []
```
Comment (jpmartinspt) [inline canonicalwebteam/blog/common_view_logic.py]: This line could be removed. If we move the following outside the `IF`, just before we do the get_articles call.
```
            featured_article_ids = [
                article["id"] for article in featured_articles
            ]
```

Comment (tbille) [reply canonicalwebteam/blog/common_view_logic.py]: liked it, coded it, pushed it

**Pull request merged**